File: src/lib/monitoring.ts
import * as Sentry from '@sentry/browser'

const REDACTED = '[REDACTED]'
const SENSITIVE_KEY_PARTS = [
    'authorization',
    'token',
    'password',
    'secret',
    'cookie',
    'api_key',
    'headers',
    'email',
]
const EMAIL_PATTERN = /[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}/gi
const BEARER_PATTERN = /(bearer\s+)[a-z0-9._-]+/gi

let monitoringInitialized = false
let windowListenersAttached = false

const DEFAULT_TRACE_TARGETS = ['localhost', '/api', import.meta.env.VITE_API_URL || '']
const NOISE_ERROR_PATTERNS = [
    /ResizeObserver loop limit exceeded/i,
    /NetworkError when attempting to fetch resource/i,
    /Failed to fetch dynamically imported module/i,
]

type ScopeLike = {
    setExtra: (key: string, value: unknown) => void
    setTag: (key: string, value: string) => void
    setLevel: (level: string) => void
}

const parseBoolean = (value: string | undefined, defaultValue: boolean): boolean => {
    if (!value) return defaultValue
    const normalized = value.trim().toLowerCase()
    if (['1', 'true', 'yes', 'on'].includes(normalized)) return true
    if (['0', 'false', 'no', 'off'].includes(normalized)) return false
    return defaultValue
}

const looksSensitive = (key: string): boolean => {
    const lowered = key.toLowerCase()
    return SENSITIVE_KEY_PARTS.some((part) => lowered.includes(part))
}

const sanitizeScalar = (value: unknown): unknown => {
    if (typeof value !== 'string') return value
    return value.replace(EMAIL_PATTERN, REDACTED).replace(BEARER_PATTERN, '$1[REDACTED]')
}

export const redactPayload = (value: unknown): unknown => {
    if (Array.isArray(value)) {
        return value.map((item) => redactPayload(item))
    }

    if (value && typeof value === 'object') {
        return Object.entries(value as Record<string, unknown>).reduce<Record<string, unknown>>((acc, [key, inner]) => {
            acc[key] = looksSensitive(key) ? REDACTED : redactPayload(inner)
            return acc
        }, {})
    }

    return sanitizeScalar(value)
}

const createBeforeSend = (event: Sentry.ErrorEvent): Sentry.ErrorEvent => {
    const scrubbed = redactPayload(event) as Sentry.ErrorEvent
    if (scrubbed.request) {
        scrubbed.request.headers = {
            redacted: REDACTED,
        }
        if (scrubbed.request.url) {
            scrubbed.request.url = scrubbed.request.url.split('?')[0]
        }
    }
    if (scrubbed.user) {
        const scrubbedUser = scrubbed.user as Record<string, unknown>
        delete scrubbedUser.email
        delete scrubbedUser.ip_address
        delete scrubbedUser.ipAddress
    }
    return scrubbed
}

const createBeforeBreadcrumb = (breadcrumb: Sentry.Breadcrumb): Sentry.Breadcrumb => {
    return redactPayload(breadcrumb) as Sentry.Breadcrumb
}

export const initMonitoring = (): boolean => {
    if (monitoringInitialized) return true

    const dsn = import.meta.env.VITE_SENTRY_DSN?.trim()
    const enabled = parseBoolean(import.meta.env.VITE_SENTRY_ENABLED, true)

    if (!enabled || !dsn) {
        return false
    }

    const tracesRate = Number.parseFloat(import.meta.env.VITE_SENTRY_TRACES_SAMPLE_RATE || '0.1')
    const release = import.meta.env.VITE_SENTRY_RELEASE?.trim()
        || import.meta.env.VITE_APP_VERSION?.trim()
        || undefined

    Sentry.init({
        dsn,
        environment: import.meta.env.MODE,
        release,
        tracesSampleRate: Number.isFinite(tracesRate) ? Math.min(Math.max(tracesRate, 0), 1) : 0.1,
        sendDefaultPii: false,
        integrations: [Sentry.browserTracingIntegration()],
        tracePropagationTargets: DEFAULT_TRACE_TARGETS.filter(Boolean),
        ignoreErrors: NOISE_ERROR_PATTERNS,
        beforeSend: createBeforeSend,
        beforeBreadcrumb: createBeforeBreadcrumb,
    })

    if (typeof window !== 'undefined' && !windowListenersAttached) {
        window.addEventListener('error', (event: ErrorEvent) => {
            if (!monitoringInitialized) return
            const error = event.error instanceof Error
                ? event.error
                : new Error(event.message || 'Unhandled window error')
            captureFrontendError(error, { source: 'window.error', message: event.message })
            if (typeof event.preventDefault === 'function') {
                event.preventDefault()
            }
        })

        window.addEventListener('unhandledrejection', (event: PromiseRejectionEvent) => {
            if (!monitoringInitialized) return
            const error = event.reason instanceof Error
                ? event.reason
                : new Error(String(event.reason || 'Unhandled promise rejection'))
            captureFrontendError(error, { source: 'window.unhandledrejection' })
            if (typeof event.preventDefault === 'function') {
                event.preventDefault()
            }
        })

        windowListenersAttached = true
    }

    monitoringInitialized = true
    return true
}

export const isMonitoringEnabled = (): boolean => {
    return monitoringInitialized
}

export const captureFrontendError = (error: Error, context?: Record<string, unknown>): void => {
    if (!monitoringInitialized) return

    const sanitizedContext = redactPayload(context || {}) as Record<string, unknown>
    Sentry.withScope((scope: ScopeLike) => {
        Object.entries(sanitizedContext).forEach(([key, value]) => {
            scope.setExtra(key, value)
        })
        Sentry.captureException(error)
    })
}

export const setMonitoringUser = (user: { id?: string | null } | null): void => {
    if (!monitoringInitialized) return
    if (!user?.id) {
        Sentry.setUser(null)
        return
    }

    Sentry.setUser({
        id: user.id || undefined,
    })
}

export const setMonitoringRoute = (route: string): void => {
    if (!monitoringInitialized || !route) return
    Sentry.setTag('route', route)
    Sentry.setContext('page', { route })
}

export const getTracePropagationHeaders = (): Record<string, string> => {
    if (!monitoringInitialized) return {}
    const traceData = Sentry.getTraceData()
    const headers: Record<string, string> = {}
    const sentryTrace = traceData['sentry-trace']
    const baggage = traceData.baggage
    if (typeof sentryTrace === 'string' && sentryTrace) {
        headers['sentry-trace'] = sentryTrace
    }
    if (typeof baggage === 'string' && baggage) {
        headers.baggage = baggage
    }
    return headers
}

export const trackTelemetry = (eventName: string, payload: Record<string, unknown> = {}): void => {
    const sanitizedPayload = redactPayload(payload) as Record<string, unknown>

    if (typeof window !== 'undefined') {
        window.dispatchEvent(new CustomEvent('kb-telemetry', {
            detail: {
                event: eventName,
                payload: sanitizedPayload,
            },
        }))
    }

    if (!monitoringInitialized) {
        return
    }

    Sentry.withScope((scope: ScopeLike) => {
        scope.setLevel('info')
        scope.setTag('telemetry_event', eventName)
        Object.entries(sanitizedPayload).forEach(([key, value]) => {
            scope.setExtra(key, value)
        })
        Sentry.captureMessage(`telemetry.${eventName}`)
    })
}

File: api/monitoring.py
"""Sentry monitoring and telemetry helpers with privacy-safe defaults."""

from __future__ import annotations

import os
import re
import hashlib
from typing import Any

from logging_config import logger

try:
    import sentry_sdk
    from sentry_sdk.integrations.fastapi import FastApiIntegration
except Exception:  # pragma: no cover - dependency may be absent in local dev
    sentry_sdk = None  # type: ignore[assignment]
    FastApiIntegration = None  # type: ignore[assignment]


_REDACTED = "[REDACTED]"
_SENSITIVE_KEY_PARTS = (
    "authorization",
    "token",
    "password",
    "secret",
    "cookie",
    "api_key",
    "headers",
    "email",
)
_EMAIL_PATTERN = re.compile(r"[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}", re.IGNORECASE)
_BEARER_PATTERN = re.compile(r"(bearer\s+)[a-z0-9._\-]+", re.IGNORECASE)
_NOISE_MESSAGES = (
    "client disconnected",
    "cancellederror",
    "stream read timed out",
)


_sentry_ready = False


def _parse_bool(value: Any, default: bool) -> bool:
    if isinstance(value, bool):
        return value
    if isinstance(value, str):
        normalized = value.strip().lower()
        if normalized in {"1", "true", "yes", "on"}:
            return True
        if normalized in {"0", "false", "no", "off"}:
            return False
    return default


def _parse_sample_rate(value: Any, default: float) -> float:
    try:
        parsed = float(value)
    except (TypeError, ValueError):
        parsed = default
    return max(0.0, min(parsed, 1.0))


def _hash_value(value: str) -> str:
    return hashlib.sha256(value.encode("utf-8")).hexdigest()[:16]


def hash_for_monitoring(value: str | None) -> str | None:
    """Return a short, non-reversible hash for monitoring contexts."""
    if not value:
        return None
    return _hash_value(value)


def _looks_sensitive(key: str) -> bool:
    lowered = key.lower()
    return any(part in lowered for part in _SENSITIVE_KEY_PARTS)


def _sanitize_scalar(value: Any) -> Any:
    if isinstance(value, str):
        scrubbed = _EMAIL_PATTERN.sub(_REDACTED, value)
        scrubbed = _BEARER_PATTERN.sub(r"\1[REDACTED]", scrubbed)
        return scrubbed
    return value


def redact_pii(value: Any) -> Any:
    """Recursively redact PII and secrets before shipping telemetry."""
    if isinstance(value, dict):
        output: dict[str, Any] = {}
        for key, inner in value.items():
            key_str = str(key)
            if _looks_sensitive(key_str):
                output[key_str] = _REDACTED
            else:
                output[key_str] = redact_pii(inner)
        return output
    if isinstance(value, list):
        return [redact_pii(item) for item in value]
    if isinstance(value, tuple):
        return tuple(redact_pii(item) for item in value)
    return _sanitize_scalar(value)


def _before_send(event: Any, _hint: Any) -> Any:
    scrubbed = redact_pii(event)
    message_value = scrubbed.get("message") if isinstance(scrubbed, dict) else None
    if isinstance(message_value, str) and any(noise in message_value.lower() for noise in _NOISE_MESSAGES):
        return None

    exception_values = (
        scrubbed.get("exception", {}).get("values", [])
        if isinstance(scrubbed, dict)
        else []
    )
    if isinstance(exception_values, list):
        for value in exception_values:
            if not isinstance(value, dict):
                continue
            exc_type = str(value.get("type") or "").lower()
            exc_value = str(value.get("value") or "").lower()
            if exc_type == "httpexception" and ("401" in exc_value or "403" in exc_value or "404" in exc_value):
                return None

    user = scrubbed.get("user")
    if isinstance(user, dict):
        user.pop("ip_address", None)
        user.pop("email", None)
    request_data = scrubbed.get("request")
    if isinstance(request_data, dict):
        request_data["headers"] = _REDACTED
        if "cookies" in request_data:
            request_data["cookies"] = _REDACTED
        url = request_data.get("url")
        if isinstance(url, str) and "?" in url:
            request_data["url"] = url.split("?", 1)[0]
    return scrubbed


def _before_breadcrumb(crumb: Any, _hint: Any) -> Any:
    return redact_pii(crumb)


def init_sentry(settings: Any) -> bool:
    """Initialize Sentry when enabled (default) and DSN is configured."""
    global _sentry_ready

    dsn = str(getattr(settings, "sentry_dsn", "") or os.getenv("SENTRY_DSN", "")).strip()
    enabled_value = getattr(settings, "sentry_enabled", os.getenv("SENTRY_ENABLED", "true"))
    enabled = _parse_bool(enabled_value, True)

    if not enabled:
        logger.info("sentry_disabled_by_env")
        _sentry_ready = False
        return False

    if not dsn:
        logger.info("sentry_disabled_missing_dsn")
        _sentry_ready = False
        return False

    if sentry_sdk is None:
        logger.warning("sentry_sdk_not_installed", dsn_present=True)
        _sentry_ready = False
        return False

    traces_rate = _parse_sample_rate(
        getattr(settings, "sentry_traces_sample_rate", os.getenv("SENTRY_TRACES_SAMPLE_RATE", 0.1)),
        0.1,
    )
    profiles_rate = _parse_sample_rate(
        getattr(settings, "sentry_profiles_sample_rate", os.getenv("SENTRY_PROFILES_SAMPLE_RATE", 0.0)),
        0.0,
    )
    environment = str(getattr(settings, "environment", "development") or "development")
    if environment == "production" and traces_rate >= 1.0:
        traces_rate = 0.2
    release = str(
        getattr(settings, "sentry_release", "")
        or os.getenv("SENTRY_RELEASE", "")
    ).strip() or None

    integrations = [FastApiIntegration()] if FastApiIntegration is not None else []
    sentry_sdk.init(
        dsn=dsn,
        environment=environment,
        release=release,
        traces_sample_rate=traces_rate,
        profiles_sample_rate=profiles_rate,
        send_default_pii=False,
        before_send=_before_send,
        before_breadcrumb=_before_breadcrumb,
        integrations=integrations,
    )
    _sentry_ready = True
    logger.info("sentry_initialized", environment=str(getattr(settings, "environment", "development")))
    return True


def sentry_is_enabled() -> bool:
    return _sentry_ready


def capture_exception(exc: Exception, **context: Any) -> None:
    """Capture backend exceptions with redacted contextual fields."""
    if not _sentry_ready or sentry_sdk is None:
        return
    with sentry_sdk.new_scope() as scope:
        for key, value in redact_pii(context).items():
            scope.set_extra(str(key), value)
        sentry_sdk.capture_exception(exc)


def capture_telemetry_event(event_name: str, **payload: Any) -> None:
    """Send low-volume telemetry events to Sentry with redaction."""
    if not _sentry_ready or sentry_sdk is None:
        return
    scrubbed = redact_pii(payload)
    with sentry_sdk.new_scope() as scope:
        scope.set_level("info")
        scope.set_tag("telemetry_event", event_name)
        for key, value in scrubbed.items():
            scope.set_extra(str(key), value)
        sentry_sdk.capture_message(f"telemetry.{event_name}")


def continue_trace_from_headers(headers: dict[str, str]) -> None:
    """Continue incoming distributed trace when propagation headers are present."""
    if not _sentry_ready or sentry_sdk is None:
        return
    sentry_trace = headers.get("sentry-trace")
    baggage = headers.get("baggage")
    if not sentry_trace:
        return
    sentry_sdk.continue_trace({"sentry-trace": sentry_trace, "baggage": baggage})


def set_request_context(*, request_id: str | None, path: str | None, method: str | None, client_ip: str | None) -> None:
    """Attach request metadata onto the active Sentry scope."""
    if not _sentry_ready or sentry_sdk is None:
        return
    scope = sentry_sdk.get_isolation_scope()
    if request_id:
        scope.set_tag("request_id", request_id)
    if path:
        scope.set_tag("path", path)
    if method:
        scope.set_tag("method", method)
    if client_ip:
        scope.set_extra("client_ip_hash", _hash_value(client_ip))


def set_user_context(
    *,
    user_id: str | None,
    email_hash: str | None = None,
    token_hash: str | None = None,
) -> None:
    """Attach authenticated user/session context to Sentry scope.

    Expects hashed inputs only; do not pass plaintext PII or secrets.
    """
    if not _sentry_ready or sentry_sdk is None:
        return

    if user_id:
        user_data: dict[str, Any] = {"id": user_id}
        if email_hash:
            user_data["email_hash"] = email_hash
        sentry_sdk.set_user(user_data)
    else:
        sentry_sdk.set_user(None)

    if token_hash:
        scope = sentry_sdk.get_isolation_scope()
        scope.set_context("session", {"token_hash": token_hash})
